src/core/net/dns_client.hpp

```diff
@@ -699,6 +730,16 @@ class Client : public InstanceLocator, private NonCopyable
 #endif
     };
 
+#if OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE
+    enum TcpState : uint8_t
+    {
+        kTcpUninitialized = 0,
+        kTcpConecting,
+        kTcpInitialized,
+        kTcpSendingInProgress,
+    };
```

Comment:
I am thinking how we tranistion between the states:
- `kTcpUninitialized` we have not yet connected.
- `kTcpConnecting` - we issue `Connect()` waiting for `HandleTcpEstablishedCallback()`, we have prepared the TCP message in `mSendLink`
- Once we get the `HandleTcpEstablishedCallback()` we pass the `mSendLInk` to TCP and enter `kTcpSendingInProgress`.
- Then we wait for `HandleSendDone()` and the enter `kTcpInitialized` (we can reclaim the `mSendLink`)
- In `kTcpInitialized` we are waiting for response to query from server, but can send another query if we need to, and can reuse the `mSendLink` again.

Does this all sound correct? 


May be better to rename and reorder the states:
- `kTcpUninitalized`
- `kTcpConnecting`
-  `kTcpConnectedSending` (instead of `kTcpSendingInProgress`)
-  `kTcpConnectedIdle`  (instead of `kTcpInitialized`)

What do you think?

Reply:
Great idea.